### ISSUE TITLE
Unexpected change in analysis results when removing `extends Object`

### ISSUE THREAD

**Issue by bwilkerson** [dart-lang/sdk#45959]
Given a file `a.dart` containing

```dart
class A {
  void _foo() {}
}

class B {
  void _foo() {}
}
```

and another file containing

```dart
import 'a.dart';

class C extends Object with A, B {}
```

the analyzer produces the following expected diagnostic

```
error: The private name '_foo', defined by 'B', conflicts with the same name defined by 'A'. (private_collision_in_mixin_application)
```

But if the extends clause is removed:

```dart
import 'a.dart';

class C extends Object with A, B {}
```

then the diagnostic is not reported.

Is this a bug in the analyzer, or is this expected behavior?

**Comment by lrhn** [comment]
If there is a difference in behavior, then it's a bug.
A class declaration of `class C with A, B {}` is completely equivalent to on of `class C extends Object with A, B {}`.

I'll admit I can't remember the special cases around mixin-application and private names (https://github.com/dart-lang/sdk/issues/28809 mentions it, but I can't find anything in the spec).
Whichever is right, both programs should behave the same.

**Comment by bwilkerson** [comment]
@eernstg For clarification on the spec.

**Comment by scheglov** [comment]
I can reproduce it.

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/199062

**Comment by eernstg** [comment]
Searching the language specification and other documents (but this is from 2017, so it should be in the language specification by now), I also don't see anything about this error. But it should definitely never make a difference whether `extends Object` is written explicitly or implied because there is no `extends` clause.

I proposed a specification of the private collision error in https://github.com/dart-lang/language/pull/1626.

I can see that we have tests, in particular `pkg/analyzer/test/src/diagnostics/private_collision_in_mixin_application_test.dart`, and they are testing the case where the colliding private declarations occur directly in the superclass and mixin, so we might not have tests covering the situation where the mixin member collides with some other concrete member that the superclass has (but doesn't declare), in which case the superclass may also be declared in some other library.

**Comment by eernstg** [comment]
Yes, I did mention #28809 in the PR dart-lang/language#1626, and [this comment](https://github.com/dart-lang/sdk/issues/28809#issuecomment-284588658) implies, if `C` is intended to be an error, that there should be a conflict on `get _x` from 'lib.dart' even though the superclass `B` is not declared in 'lib.dart'.
